helper/jsonutil/json.go

```diff
@@ -38,22 +38,46 @@ func EncodeJSONAndCompress(in interface{}) ([]byte, error) {
 		return nil, err
 	}
 
+	if config == nil {
+		config = &compressutil.CompressionConfig{
+			Type:                 compressutil.CompressionTypeGzip,
+			GzipCompressionLevel: gzip.BestCompression,
+		}
+	}
+
 	// For compression, use Gzip format with 'BestCompression' level.
-	return compressutil.Compress(encodedBytes, &compressutil.CompressionConfig{
-		Type:                 compressutil.CompressionTypeGzip,
-		GzipCompressionLevel: gzip.BestCompression,
-	})
+	return compressutil.Compress(encodedBytes, config)
 }
 
-// Decodes/Unmarshals the given JSON into a desired object
+// DecompressAndDecodeJSON tries to decompress the given data. The call to
+// decompress, fails if the content was not compressed in the first place,
+// which is identified by a canary byte before the compressed data. If the data
+// is not compressed, it is JSON decoded directly. Otherwise the decompressed
+// data will be JSON decoded.
 func DecodeJSON(data []byte, out interface{}) error {
-	if data == nil {
+	if data == nil || len(data) == 0 {
 		return fmt.Errorf("'data' being decoded is nil")
 	}
 	if out == nil {
 		return fmt.Errorf("output parameter 'out' is nil")
 	}
 
+	// Decompress the dataBytes using Gzip format.
```

Comment:
I think this is not yet finished? This will decompress *everything* rather than using the canary.